### PR TITLE
feat(oracle): provide balance change of native-restaking asset

### PR DESCRIPTION
**Pull request by leonz789**
## Description

This PR works cooperating with #183 , and replace #170 
Use 'amount change' instead of 'virtual price change' to reflect balance change of staked-eth on beacon chain by slash/refund.

----

Closes #XXX

## Summary by CodeRabbit

- **New Features**
	- Introduced support for managing native tokens, including functionalities for deposits, withdrawals, and balance tracking.
	- Added new Protocol Buffers messages for native token management, enhancing data representation.
	- Implemented new gRPC methods for querying staker data related to specific assets.
	- Enhanced functionality for retrieving staker information and lists through new queries.

- **Bug Fixes**
	- Improved error handling for updating native token prices.

- **Documentation**
	- Enhanced error reporting capabilities with new error constants related to native tokens.

- **Refactor**
	- Streamlined key generation for native tokens to improve data handling and access.

### CODE REVIEW

File: proto/exocore/oracle/v1/native_token.proto [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	ACTION_SLASH_REFUND = 3;
+  }
+
+  Action change = 5;
+}
+
+message StakerInfo {
+  string staker_addr = 1;
+  int64 staker_index = 2;
+  repeated uint64 validator_indexs = 3;
```
Comment (bwhour) [inline proto/exocore/oracle/v1/native_token.proto]: typo: validator_**indexes**

File: x/oracle/keeper/native_token.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	key := types.NativeTokenStakerKey(assetID, stakerAddr)
+	stakerInfo := &types.StakerInfo{}
+	if value := store.Get(key); value == nil {
+		// create a new item for this staker
+		stakerInfo = types.NewStakerInfo(stakerAddr, validatorIndex)
+	} else {
+		k.cdc.MustUnmarshal(value, stakerInfo)
+		if amount.IsPositive() {
+			// deopsit add a new validator into staker's validatorList
+			stakerInfo.ValidatorIndexs = append(stakerInfo.ValidatorIndexs, validatorIndex)
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go]: ValidatorIndexes

File: x/oracle/keeper/native_token.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+func (k Keeper) getStakerList(ctx sdk.Context, assetID string) types.StakerList {
+	if len(stakerList.StakerAddrs) == 0 {
+		stakerList = k.GetStakerList(ctx, assetID)
+	}
+	return stakerList
+}
+
+// parseBalanceChange parses rawData to details of amount change for all stakers relative to native restaking
+func parseBalanceChange(rawData []byte, sl types.StakerList) (map[string]int, error) {
+	indexs := rawData[:32]
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go]: can you add some comment for this logic ?

File: x/assets/keeper/bank.go [hunk 4]
```diff
@@ -43,6 +45,12 @@ func (k Keeper) PerformDepositOrWithdraw(ctx sdk.Context, params *DepositWithdra
 		return errorsmod.Wrapf(assetstypes.ErrInvalidOperationType, "the operation type is: %v", params.Action)
 	}
 
+	if assetstypes.IsNativeToken(assetID) {
+		// TODO: we skip check for case like withdraw amount>withdrawable is fine since it will fail for later check and the state will be rollback
+		// TODO: need validatorIndex to be passed (0 as placeholder)
+		actualOpAmount = k.UpdateNativeTokenByDepositOrWithdraw(ctx, assetID, hexutil.Encode(params.StakerAddress), params.OpAmount, 0)
```
Comment (TimmyExogenous) [inline x/assets/keeper/bank.go]: Is this used to update client chain validator information? If not, is it necessary?

File: x/assets/types/general.go [hunk 5]
```diff
@@ -27,8 +27,16 @@ const (
 	MaxChainTokenMetaInfoLength = 200
 
 	MinClientChainAddrLength = 20
+
+	// TODO: update before merge
+	NativeETHAssetID = "0x01_0x01"
```
Comment (TimmyExogenous) [inline x/assets/types/general.go:40]: If we use `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` to represent all native restaking tokens, this hardcode can be removed and we just need to provide a function to check if all bytes are `0xee`, right?

File: x/operator/keeper/usd_value.go [hunk 6]
```diff
@@ -299,6 +300,9 @@ func (k *Keeper) CalculateUSDValueForOperator(
 		if isForSlash {
 			// when calculated the USD value for slashing, the input prices map is null
 			// so the price needs to be retrieved here
+			if assetstype.IsNativeToken(assetID) {
+				assetID = strings.Join([]string{assetID, operator}, "_")
```
Comment (TimmyExogenous) [inline x/operator/keeper/usd_value.go]: Is this still needed? Seems like the native token can be considered as the base token when fetching price even if different operators own it.

File: x/oracle/keeper/native_token.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+	stakerInfo := types.StakerInfo{}
+	value := store.Get(types.NativeTokenStakerKey(assetID, stakerAddr))
+	if value == nil {
+		return stakerInfo
+	}
+	k.cdc.MustUnmarshal(value, &stakerInfo)
+	return stakerInfo
+}
+
+// TODO, NOTE: price changes will effect reward/slash calculation, every time one staker's price changed, it's reward/slash amount(LST) should be cleaned or recalculated immediately
```
Comment (TimmyExogenous) [inline x/oracle/keeper/native_token.go]: Are these code comments deprecated?

File: x/assets/keeper/bank.go (hunk 4 above)
Comment (TimmyExogenous) [reply x/assets/keeper/bank.go]: And is it an interface provided by Oracle keeper?

File: x/oracle/keeper/native_token.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	if newBalance.Balance <= 0 {
+		store.Delete(key)
+	} else {
+		bz := k.cdc.MustMarshal(stakerInfo)
+		store.Set(key, bz)
+	}
+	return amount
+}
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: This function is invoked through assets module, and the validator should have been done properly.

File: x/oracle/module.go [hunk 9]
```diff
@@ -162,6 +163,12 @@ func (am AppModule) BeginBlock(ctx sdk.Context, _ abci.RequestBeginBlock) {
 
 // EndBlock contains the logic that is automatically triggered at the end of each block
 func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
+	// debug: this is for test only to trigger this event every block
+	eventValue := fmt.Sprintf("%s_%s_%s", types.AttributeValueNativeTokenWithdraw, "0xabcefz", strconv.FormatUint(1977, 10))
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeCreatePrice,
+		sdk.NewAttribute(types.AttributeKeyNativeTokenChange, eventValue),
+	))
```
[the remark this answers is not in the document]
Comment (leonz789) [reply x/oracle/module.go]: removed

File: x/assets/keeper/bank.go (hunk 4 above)
Comment (leonz789) [reply x/assets/keeper/bank.go]: 1. no this is used to tell oracle Keeper about the validator list change for one staker
2. yes, this is an interface provided by Oracle, i embedded oracle keeper as anonymity, so i don' t need to call it as k.oraclek.UpdateNativeTokenByDepositOrwithdraw

File: x/assets/types/general.go (hunk 5 above)
Comment (leonz789) [reply x/assets/types/general.go:40]: By 'all native restaking tokens', do you mean we are not able to distinguish different assets (for native-restakign) ?

File: x/oracle/keeper/native_token.go (hunk 2 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go (hunk 3 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/oracle/keeper/native_token.go (hunk 7 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go]: updated

File: x/operator/keeper/usd_value.go (hunk 6 above)
Comment (leonz789) [reply x/operator/keeper/usd_value.go]: removed

File: proto/exocore/oracle/v1/native_token.proto (hunk 1 above)
Comment (leonz789) [reply proto/exocore/oracle/v1/native_token.proto]: updated

File: proto/exocore/oracle/v1/native_token.proto [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+  Action change = 5;
+}
+
+// StakerInfo represents all related information for a staker of native-restaking
+message StakerInfo {
+  string staker_addr = 1;
+  int64 staker_index = 2;
+  repeated uint64 validator_indexes = 3;
+  repeated BalanceInfo balance_list = 5;
```
Comment (bwhour) [inline proto/exocore/oracle/v1/native_token.proto]: no 4 ?

Comment (leonz789) [reply proto/exocore/oracle/v1/native_token.proto]: updated

File: x/oracle/keeper/prices.go [hunk 11]
```diff
@@ -187,13 +188,32 @@ func (k Keeper) AppendPriceTR(ctx sdk.Context, tokenID uint64, priceTR types.Pri
...
+	// update for native tokens
+	// TODO: set hooks as a genral approach
+	var p types.Params
+	// get params from cache if exists
+	if agc != nil {
+		p = agc.GetParams()
+	} else {
+		p = k.GetParams(ctx)
+	}
+	assetIDs := p.GetAssetIDsFromTokenID(tokenID)
```
Comment (bwhour) [inline x/oracle/keeper/prices.go:218]: if assetIDs is nil ?

Comment (leonz789) [reply x/oracle/keeper/prices.go:218]: then range will just be skipped in line_203

File: x/oracle/keeper/native_token_test.go [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+	// TODO: update after assets module got the definition
+	assetID := "0xe_" + chainID
+	validators := []string{"0xv1", "0xv2"}
+	// 1. deposit amount 100
+	// 100 is not a possible nubmer with one validator, it's ok to use this as a start and we'll check the number to be updated to a right number(uncer 32 with one validator)
+	amount100 := sdkmath.NewIntFromUint64(100)
+	amount32 := sdkmath.NewIntFromUint64(32)
+	ks.k.UpdateNativeTokenValidatorListForStaker(ks.ctx, chainID, stakerStr, validators[0], amount100)
+	// - 1.1 check stakerInfo
+	stakerInfo := ks.k.GetStakerInfo(ks.ctx, assetID, stakerStr)
```
Comment (bwhour) [inline x/oracle/keeper/native_token_test.go:51]: no case when GetStakerInfo is nil value return ?

File: x/oracle/keeper/native_token.go [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+		sInfo.BalanceList = sInfo.BalanceList[:len(sInfo.BalanceList)-1]
+		// this is mainly used by price feeder, so we remove the stakerAddr to reduce the size of return value
+		sInfo.StakerAddr = ""
+		ret = append(ret, &sInfo)
+	}
+	return ret
+}
+
+// GetstakerList return stakerList for native-restaking asset of assetID
+func (k Keeper) GetStakerList(ctx sdk.Context, assetID string) types.StakerList {
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go:151]: Consider returning an additional boolean value in the **GetStakerInfo** and **GetStakerList** functions to indicate whether the data has been found

> func (k Keeper) GetStakerInfo(ctx sdk.Context, assetID, stakerAddr string) (types.StakerInfo, bool) {
    // ...
    if value == nil {
        return stakerInfo, false
    }
    k.cdc.MustUnmarshal(value, &stakerInfo)
    return stakerInfo, true
}

File: x/oracle/keeper/native_token.go [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+
+var stakerList types.StakerList
+
+// GetStakerInfo returns details about staker for native-restaking under asset of assetID
+func (k Keeper) GetStakerInfo(ctx sdk.Context, assetID, stakerAddr string) types.StakerInfo {
+	store := ctx.KVStore(k.storeKey)
+	stakerInfo := types.StakerInfo{}
+	value := store.Get(types.NativeTokenStakerKey(assetID, stakerAddr))
+	if value == nil {
+		return stakerInfo
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go:28]: Consider returning an additional boolean value in the **GetStakerInfo** and **GetStakerList** functions to indicate whether the data has been found

> func (k Keeper) GetStakerInfo(ctx sdk.Context, assetID, stakerAddr string) (types.StakerInfo, bool) {
    // ...
    if value == nil {
        return stakerInfo, false
    }
    k.cdc.MustUnmarshal(value, &stakerInfo)
    return stakerInfo, true
}

File: x/oracle/keeper/native_token_test.go (hunk 12 above)
Comment (leonz789) [reply x/oracle/keeper/native_token_test.go:51]: line129-130 is nil case: the staker has been removed and return value is empty struct( when nil)

File: x/oracle/keeper/native_token.go (hunk 13 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:151]: when not found, it will return empty struct

File: x/oracle/keeper/native_token.go (hunk 14 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:28]: when not found, it will return empty struct

File: x/oracle/keeper/native_token.go (hunk 13 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:151]: stakerList is a wrapper of slice, so could be checked with length

File: x/oracle/keeper/native_token.go (hunk 14 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:28]: stakerList is a wrapper of slice, so could be checked with length

File: x/oracle/keeper/query_native_token.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (k Keeper) StakerInfos(goCtx context.Context, req *types.QueryStakerInfosRequest) (*types.QueryStakerInfosResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request for stakerInfos")
+	}
+	if !assetstypes.IsNativeToken(req.AssetID) {
+		return nil, errors.New("assetID doesn't reprensents any supported nativeRestakingToken")
```
Comment (bwhour) [inline x/oracle/keeper/query_native_token.go]: the error can be var like the following:
>var (
    ErrInvalidRequest = status.Error(codes.InvalidArgument, "invalid request")
    ErrUnsupportedAsset = errors.New("assetID doesn't represent any supported native restaking token")
)

File: x/oracle/keeper/query_native_token.go [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+	}
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	stakerInfos := k.GetStakerInfos(ctx, req.AssetID)
+	return &types.QueryStakerInfosResponse{StakerInfos: stakerInfos}, nil
+}
+
+func (k Keeper) StakerList(goCtx context.Context, req *types.QueryStakerListRequest) (*types.QueryStakerListResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request for stakerList")
```
Comment (bwhour) [inline x/oracle/keeper/query_native_token.go]: the error can be var like the following:
>var (
    ErrInvalidRequest = status.Error(codes.InvalidArgument, "invalid request")
    ErrUnsupportedAsset = errors.New("assetID doesn't represent any supported native restaking token")
)

File: x/oracle/types/key_native_token.go [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+func NativeTokenStakerKeyPrefix(assetID string) []byte {
+	return append([]byte(NativeTokenStakerInfoKeyPrefix), []byte(assetID)...)
+}
+
+func NativeTokenOperatorKey(assetID, operatorAddr string) []byte {
+	assetID = strings.Join([]string{assetID, operatorAddr}, "/")
+	return append([]byte(NativeTokenOperatorInfoKeyPrefix), []byte(assetID)...)
+}
+
+// func NativeTokenValidatorListKey(assetID string) []byte {
```
Comment (bwhour) [inline x/oracle/types/key_native_token.go]: remove this

File: x/oracle/types/key_native_token.go [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+const (
+	// NativeTokenKeyPrefix is the prefix to retrieve all NativeToken
+	NativeTokenKeyPrefix                 = "NativeToken/"
+	NativeTokenPriceKeyPrefix            = NativeTokenKeyPrefix + "price/value/"
+	NativeTokenStakerInfoKeyPrefix       = NativeTokenKeyPrefix + "stakerInfo/value/"
+	NativeTokenOperatorInfoKeyPrefix     = NativeTokenKeyPrefix + "operatorInfo/value/"
+	NativeTokenStakerListKeyPrefix       = NativeTokenKeyPrefix + "stakerList/value/"
+	NativeTokenStakerDelegationKeyPrefix = NativeTokenKeyPrefix + "stakerDelegation/value/"
+
+// NativeTokenValidatorListKeyPrefix    = NativeTokenKeyPrefix + "validatorList/value"
```
Comment (bwhour) [inline x/oracle/types/key_native_token.go]: remove this

File: x/oracle/types/key_native_token.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+	assetID = strings.Join([]string{assetID, operatorAddr}, "/")
+	return append([]byte(NativeTokenOperatorInfoKeyPrefix), []byte(assetID)...)
+}
+
+// func NativeTokenValidatorListKey(assetID string) []byte {
+// 	return append([]byte(NativeTokenValidatorListKeyPrefix), []byte(assetID)...)
+// }
+
+func ParseNativeTokenStakerKey(key []byte) (assetID, stakerAddr string) {
+	parsed := strings.Split(string(key), "/")
```
Comment (bwhour) [inline x/oracle/types/key_native_token.go:39]: Use SplitN to limit the number of divisions

>parsed := **strings.SplitN**(string(key), "/", 4)

File: x/oracle/types/key_native_token.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+	assetID = strings.Join([]string{assetID, stakerAddr}, "/")
+	return append([]byte(NativeTokenStakerDelegationKeyPrefix), []byte(assetID)...)
+}
+
+func NativeTokenStakerListKey(assetID string) []byte {
+	return append([]byte(NativeTokenStakerListKeyPrefix), []byte(assetID)...)
+}
+
+func NativeTokenStakerKey(assetID, stakerAddr string) []byte {
+	assetID = strings.Join([]string{assetID, stakerAddr}, "/")
```
Comment (bwhour) [inline x/oracle/types/key_native_token.go:25]: no need check the params ?
> if assetID == "" || stakerAddr == "" {
        panic("assetID and stakerAddr must not be empty")
    }

File: x/oracle/types/key_native_token.go [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+// NativeTokenValidatorListKeyPrefix    = NativeTokenKeyPrefix + "validatorList/value"
+)
+
+func NativeTokenStakerDelegationKey(assetID, stakerAddr string) []byte {
+	assetID = strings.Join([]string{assetID, stakerAddr}, "/")
+	return append([]byte(NativeTokenStakerDelegationKeyPrefix), []byte(assetID)...)
+}
+
+func NativeTokenStakerListKey(assetID string) []byte {
```
Comment (bwhour) [inline x/oracle/types/key_native_token.go:20]: If some function is frequently called on a performance-critical path, consider using sync.Pool to reuse bytes.Buffer

File: x/oracle/keeper/native_token.go [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+		}
+		//	newBalance.Balance += int64(change)
+		stakerInfo.Append(&newBalance)
+		bz := k.cdc.MustMarshal(stakerInfo)
+		store.Set(key, bz)
+	}
+	return nil
+}
+
+// getStakerList returns all Stakers for native-restaking of assetID, this is used for cache
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go:217]: GetstakerList -> GetStakerList

File: x/oracle/keeper/native_token.go [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+	newBalance := types.BalanceInfo{}
+
+	if latestIndex := len(stakerInfo.BalanceList) - 1; latestIndex >= 0 {
+		newBalance = *(stakerInfo.BalanceList[latestIndex])
+		newBalance.Index++
+	}
+	//	newBalance := *(stakerInfo.BalanceList[latestIndex])
+	//	newBalance.Index++
+	newBalance.Block = uint64(ctx.BlockHeight())
+	if amount.IsPositive() {
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go:63]: Is there a validity check for the **amount** value range?

File: x/oracle/keeper/native_token.go [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+	return nil
+}
+
+// GetStakerInfos returns all stakers information
+func (k Keeper) GetStakerInfos(ctx sdk.Context, assetID string) (ret []*types.StakerInfo) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.NativeTokenStakerKeyPrefix(assetID))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		sInfo := types.StakerInfo{}
```
Comment (bwhour) [inline x/oracle/keeper/native_token.go:139]: Reuse objects StakerInfo  using sync.Pool ?

File: x/oracle/keeper/prices.go (hunk 11 above)
Comment (bwhour) [reply x/oracle/keeper/prices.go:218]: ok,make sure it doesn't affect the logic.

File: x/oracle/types/native_token.go [hunk 25]
```diff
@@ -0,0 +1,21 @@
+package types
+
+const maxSize = 100
+
+// TODO: vlaidatorIndex need bridge data
```
Comment (bwhour) [inline x/oracle/types/native_token.go]: no index, update comment here.

File: x/oracle/types/key_native_token.go (hunk 19 above)
Comment (leonz789) [reply x/oracle/types/key_native_token.go:39]: SplitN will split string into slice with lenght=N and keep the remaining as a single string, that is not expectec.
We've got check `if len(parsed)!=3` ....

File: x/oracle/types/key_native_token.go (hunk 21 above)
Comment (leonz789) [reply x/oracle/types/key_native_token.go:20]: not frequently called

File: x/oracle/keeper/native_token.go (hunk 23 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:63]: this is called through from assets module, and assets module is responsible for the validity check

File: x/oracle/keeper/native_token.go (hunk 24 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:139]: sInfo will be appended to slice as pointer, So the memory should be allocated anyway(need to do deep copy if we use sync.pool).

File: x/oracle/types/key_native_token.go (hunk 21 above)
Comment (leonz789) [reply x/oracle/types/key_native_token.go:22]: not frequently

File: x/oracle/types/native_token.go (hunk 25 above)
Comment (leonz789) [reply x/oracle/types/native_token.go]: updated

File: x/oracle/keeper/native_token.go (hunk 22 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:217]: updated

File: x/oracle/types/key_native_token.go (hunk 18 above)
Comment (leonz789) [reply x/oracle/types/key_native_token.go]: updated

File: x/oracle/types/key_native_token.go (hunk 17 above)
Comment (leonz789) [reply x/oracle/types/key_native_token.go]: updated

File: x/oracle/keeper/query_native_token.go (hunk 16 above)
Comment (leonz789) [reply x/oracle/keeper/query_native_token.go]: updated

File: x/oracle/keeper/query_native_token.go (hunk 15 above)
Comment (leonz789) [reply x/oracle/keeper/query_native_token.go]: updated

File: proto/exocore/oracle/v1/native_token.proto [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+  // on which block the balance updated, deposit/withdraw will update this and round_id might not change
+  uint64 block = 2;
+  // index used to distinguish multiple balanceInfo for one same roundid
+  uint64 index = 3;
+  // beaconchain epoch related to slash_refund
+  // uint epoch = 3;
+  int64 balance = 4;
+
+  // Action defines the reason the balance changed
+  enum Action {
```
Comment (MaxMustermann2) [inline proto/exocore/oracle/v1/native_token.proto]: nit: move this outside `BalanceInfo` please. not a blocker.

File: x/assets/keeper/bank.go [hunk 27]
```diff
@@ -36,6 +38,14 @@ func (k Keeper) PerformDepositOrWithdraw(ctx sdk.Context, params *DepositWithdra
 		return errorsmod.Wrapf(assetstypes.ErrInvalidOperationType, "the operation type is: %v", params.Action)
 	}
 
+	if assetstypes.IsNativeToken(assetID) {
```
Comment (MaxMustermann2) [inline x/assets/keeper/bank.go]: This is a bit confusing. We should use `NST` to distinguish with native tokens from #127 or maybe say native restaked token?

File: x/oracle/client/cli/query_native_token.go [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+
+			queryClient := types.NewQueryClient(clientCtx)
+
+			assetID := args[0]
```
Comment (MaxMustermann2) [inline x/oracle/client/cli/query_native_token.go:24]: Stateless validate this via `assetstypes.ValidateID` ? Not a blocker.

File: x/oracle/client/cli/query_native_token.go [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+
+			queryClient := types.NewQueryClient(clientCtx)
+
+			request := &types.QueryStakerInfoRequest{
+				AssetId:    args[0],
```
Comment (MaxMustermann2) [inline x/oracle/client/cli/query_native_token.go]: Stateless validate this via `assetstypes.ValidateID` ? Not a blocker.

File: x/oracle/keeper/native_token.go [hunk 30]
```diff
@@ -0,0 +1,292 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// deposit: update staker's totalDeposit
+// withdoraw: update staker's totalDeposit
+// delegate: update operator's price, operator's totalAmount, operator's totalShare, staker's share
+// undelegate: update operator's price, operator's totalAmount, operator's totalShare, staker's share
+// msg(refund or slash on beaconChain): update staker's price, operator's price
+
+const maxEffectiveBalance = 32
```
Comment (MaxMustermann2) [inline x/oracle/keeper/native_token.go:18]: This is only for ETH right?

**Review by MaxMustermann2 — APPROVED** [review]

File: x/oracle/keeper/native_token.go (hunk 30 above)
Comment (leonz789) [reply x/oracle/keeper/native_token.go:18]: yes

File: proto/exocore/oracle/v1/native_token.proto (hunk 26 above)
Comment (leonz789) [reply proto/exocore/oracle/v1/native_token.proto]: updated

File: x/assets/keeper/bank.go (hunk 27 above)
Comment (leonz789) [reply x/assets/keeper/bank.go]: updated

File: x/oracle/client/cli/query_native_token.go (hunk 29 above)
Comment (leonz789) [reply x/oracle/client/cli/query_native_token.go]: updated

**Review by TimmyExogenous — APPROVED** [review]

File: x/oracle/client/cli/query_native_token.go (hunk 28 above)
Comment (leonz789) [reply x/oracle/client/cli/query_native_token.go:24]: updated

**Review by bwhour — APPROVED** [review]

**Pull request merged**